app/controllers/application_controller/miq_request_methods.rb

```diff
@@ -984,69 +984,37 @@ def workflow_instance_from_vars(req)
-        end
-        @curr_tag = t[:name]
-        @ci_node = {} # Build the ci node
-        @ci_node[:key] = t[:id].to_s
-        @ci_node[:title] = t[:description]
-        @ci_node[:title] += " *" if t[:single_value]
-        @ci_node[:tooltip] = t[:description]
-        @ci_node[:addClass] = "cfme-no-cursor-node" # No cursor pointer
-        @ci_node[:icon] = 'pficon pficon-folder-close'
-        @ci_node[:hideCheckbox] = @ci_node[:cfmeNoClick] = true
-        @ci_node[:addClass] = "cfme-bold-node" # Show node as different
-        @ci_kids = []
-      end
-      if @curr_tag.present? && @curr_tag == t[:name]
-        t[:children].each do |c|
-          temp = {}
-          temp[:key] = c[0].to_s
-          # only add cfme_parent_key for single value tags, need to use in JS onclick handler
-          temp[:selectable] = false
-          temp[:title] = temp[:tooltip] = c[1][:description]
-          temp[:addClass] = "cfme-no-cursor-node"
-          temp[:icon] = 'fa fa-tag'
-          if edit_mode # Don't show checkboxes/radio buttons in non-edit mode
-            if vm_tags && vm_tags.include?(c[0].to_i)
-              temp[:select] = true
-              kids_checked = true
-            else
-              temp[:select] = false
-            end
-            if @edit && @edit[:current][tag_symbol_for_workflow] != @edit[:new][tag_symbol_for_workflow]
-              # checking to see if id is in current but not in new, change them to blue OR if id is in current but deleted from new
-              if (!@edit[:current][tag_symbol_for_workflow].include?(c[0].to_i) && @edit[:new][tag_symbol_for_workflow].include?(c[0].to_i)) ||
-                 (!@edit[:new][tag_symbol_for_workflow].include?(c[0].to_i) && @edit[:current][tag_symbol_for_workflow].include?(c[0].to_i))
-                temp[:addClass] = "cfme-blue-bold-node"
-              end
-            end
-            @ci_kids.push(temp) unless @ci_kids.include?(temp)
-          else
-            temp[:hideCheckbox] = true
-            @ci_kids.push(temp) unless @ci_kids.include?(temp) || !vm_tags.include?(c[0].to_i)
-          end
-        end
-      end
-      if i == tags.length - 1 # Adding last node
-        @ci_node[:expand] = true if kids_checked
-        kids_checked = false
-        @ci_node[:children] = @ci_kids if @ci_kids.present?
-        all_tags.push(@ci_node) if @ci_kids.present?
+    # for some reason @tags is set in wf, and it is changed by map bellow which causes bugs
+    wf.instance_variable_set(:@tags, nil)
+    tags = wf.allowed_tags.map do |cat|
+      cat[:values] = cat[:children].dup.map do |tag|
+        { :id => tag.first, :description => tag.second[:description] }
       end
+      cat.delete(:children)
+      cat
     end
-    @all_tags_tree = TreeBuilder.convert_bs_tree(all_tags).to_json # Add ci node array to root of tree
+    tags = tags.map { |t| {:id => t[:name], :singleValue => t[:single_value] }.merge(t) }
```

Comment:
You have better hash manipulation techniques for this, check out the [documentation](https://ruby-doc.org/core-2.5.1/Hash.html)